File: DS/FenwickTree.py
class FenwickTree:
    def __init__(self, n):
        self.tree = [0]*(n+1)
    def update(self, i, diff):
        i += 1
        while i < len(self.tree):
            self.tree[i] += diff
            i += i & (-i)
    def _rangeSum(self, i):
        i += 1
        sum = 0
        while i > 0:
            sum += self.tree[i]
            i -= i & (-i)
        return sum
    def rangeSum(self, i, j):
        return self._rangeSum(j) - self._rangeSum(i-1)

if __name__ == '__main__':
    m, n = map(int, input().split())
    nums = list(map(int, input().split()))
    ft = FenwickTree(nums)
    for x in range(m):
        ft.update(x, nums[x])
    for i in range(n):
        a, b, c = map(int, input().split())
        if a == 1:
            ft.update(b-1, c-nums[b-1])
            nums[b-1] = c
        else:
            print(ft.rangeSum(b-1, c-1))

File: DS/FenwickTree.java
package DS;

// Similar to Segment Tree, but with a different implementation and more efficient for range sum queries
public class FenwickTree {
    int[] tree;
    FenwickTree(int[] nums){
        tree = new int[nums.length+1];
    }
    void update(int i,int diff){
        i++; // need position rather than index
        while (i<tree.length) {
            tree[i] +=diff;
            i += (i&(-i)); // get next position to update by removing last 1 bit 110 -> 10 or 11010100 -> 100
        }
    }
    int rangeSum(int i){
        i++;
        int sum=0;
        while (i>0) {
            sum += tree[i];
            i -= (i&(-i)); // maintain the last 1 bit
        }
        return sum;
    }
    int rangeSum(int i,int j){
        return rangeSum(j)-rangeSum(i-1);
    }
    public static void main(String[] args) {
        int[] nums = {1,2,3,4,5,6,7,8,9,10};
        FenwickTree ft = new FenwickTree(nums);
        for (int i = 0; i < nums.length; i++) {
            ft.update(i, nums[i]);
        }
        System.out.println(ft.rangeSum(2, 2));
    }
}
